waiter/src/waiter/descriptor.clj

```diff
@@ -0,0 +1,268 @@
+      (when (and (pos? service-fallback-period-secs)
+                 (let [most-recently-modified-token (sd/retrieve-most-recently-modified-token token->token-data)
+                       token-last-update-time (get-in token->token-data [most-recently-modified-token "last-update-time"] 0)]
+                   (->> (t/seconds service-fallback-period-secs)
+                        (t/plus (tc/from-long token-last-update-time))
+                        (t/before? request-time))))
+        (loop [iteration 1
+               loop-descriptor descriptor]
+          (when (<= iteration search-history-length)
+            (when-let [previous-descriptor (descriptor->previous-descriptor loop-descriptor)]
+              (let [{:keys [service-id]} previous-descriptor]
+                (if (service-healthy? fallback-state service-id)
+                  (do
+                    (log/info (str "iteration-" iteration) (:service-id descriptor) "falling back to" service-id)
+                    previous-descriptor)
+                  (do
+                    (log/debug (str "iteration-" iteration) "skipping" service-id "as the fallback service"
+                               {:available (service-exists? fallback-state service-id)
+                                :healthy (service-healthy? fallback-state service-id)})
+                    (recur (inc iteration) previous-descriptor)))))))))))
+
+(defn resolve-descriptor
+  "Resolves the descriptor that should be used based on available healthy services.
+   Returns the provided latest-descriptor if it references a service with healthy instances or no valid fallback
+   descriptor can be found (i.e. no descriptor in the search history which has healthy instances).
+   Resolves to a different fallback descriptor which describes a services with healthy instances when the latest descriptor
+   represents a service that has not started or does not have healthy instances."
+  [descriptor->previous-descriptor search-history-length request-time fallback-state latest-descriptor]
+  (let [{:keys [service-id]} latest-descriptor]
+    (if (service-healthy? fallback-state service-id)
+      latest-descriptor
+      (or (retrieve-fallback-descriptor
+            descriptor->previous-descriptor search-history-length fallback-state request-time latest-descriptor)
+          (do
+            (log/info "no fallback service found for" (:service-id latest-descriptor))
+            latest-descriptor)))))
+
+(defn request-authorized?
+  "Takes the request w/ kerberos auth info & the app headers, and returns true if the user is allowed to use "
+  [user permitted-user]
+  (log/debug "validating:" (str "permitted=" permitted-user) (str "actual=" user))
+  (or (= token/ANY-USER permitted-user)
+      (= ":any" (str permitted-user)) ; support ":any" for backwards compatibility
+      (and (not (nil? permitted-user)) (= user permitted-user))))
+
+(defn compute-descriptor
```

Comment:
Renamed from `request->descriptor` to `compute-descriptor`. Moved from `service_description.clj`